csrc/scheduler/reduction.cpp

```diff
@@ -912,6 +912,22 @@ std::unique_ptr<ReductionParams> heuristicParaToSchedulerPara(
   // bool flip_grid = gidim > 1 && gidim < 8;
   const bool flip_grid = false;
   auto rparams = std::make_unique<ReductionParams>();
+
+  // Configure register count to achieve 50% occupancy with 1024 threads per SM.
+  // Typical thread configurations are 512 or 1024 threads per block, both
+  // evenly divisible. This setup ensures each thread is limited to 64
+  // registers, preventing compilers from allocating more registers per thread,
+  // which could reduce occupancy and degrade performance.
+  int64_t target_threads_per_sm =
+      (int64_t)at::cuda::getCurrentDeviceProperties()
+          ->maxThreadsPerMultiProcessor /
+      2;
+  int64_t threads_per_block = params.bdimx * params.bdimy;
+  int64_t threads_per_sm =
+      target_threads_per_sm / threads_per_block * threads_per_block;
```

Comment:
In the unlikely event where `threads_per_sm < threads_per_block`, this would give us `0`.

Reply:
revised to `scheduler_utils::safeDiv`